storage/writer.go

```diff
@@ -84,30 +94,301 @@ type Writer struct {
+		data := buf[sent : sent+limit]
+		req := &storagepb.WriteObjectRequest{
+			Data: &storagepb.WriteObjectRequest_ChecksummedData{
+				ChecksummedData: &storagepb.ChecksummedData{
+					Content: data,
+				},
+			},
+			WriteOffset: offset,
+			FinishWrite: finished,
+		}
+
+		n, err := w.sendWithRetry(req, first, upid)
+		if err != nil {
+			return nil, 0, false, err
+		}
+		first = false
+
+		// Update the immediate stream's sent total and the global upload
+		// offset with the data sent.
+		sent += n
+		offset += int64(n)
+	}
+
+	// Close the stream to "commit" the data sent.
+	resp, finalized, err := w.commit()
+	if err != nil {
+		return nil, 0, false, err
+	}
+
+	return resp.GetResource(), offset, finalized, err
+}
+
+// commit closes the stream to commit the data sent and potentially receive
+// the finalized object if finished uploading. If the last request sent
+// indicated that writing was finished, the Object will be finalized and
+// returned. If not, then the Object will be nil, and the boolean returned will
+// be false.
+func (w *Writer) commit() (*storagepb.WriteObjectResponse, bool, error) {
+	finalized := true
+	resp, err := w.stream.CloseAndRecv()
+	if err == io.EOF {
+		// Closing a stream for a resumable upload finish_write = false results
+		// in an EOF which can be ignored, as we aren't done uploading yet.
+		finalized = false
+		err = nil
+	}
+	// Drop the stream reference as it has been closed.
+	w.stream = nil
+
+	return resp, finalized, err
+}
+
+// sendWithRetry will attempt to Send a WriteObjectRequest on the stream, and
+// will reopen the stream and retry that send if it fails for a retryable
+// reason until the Writer's context is canceled.
+func (w *Writer) sendWithRetry(req *storagepb.WriteObjectRequest, first bool, upid string) (int, error) {
+	var err error
+	err = runWithRetry(w.ctx, func() error {
+		// Open a new stream and set the first_message field on the request.
+		// The first message on the WriteObject stream must either be the
+		// Object or the Resumable Upload ID.
+		if first {
+			w.stream, err = w.o.c.gc.WriteObject(w.ctx)
+			if err != nil {
+				return err
+			}
+
+			if upid != "" {
+				req.FirstMessage = &storagepb.WriteObjectRequest_UploadId{UploadId: upid}
+			} else {
+				req.FirstMessage = &storagepb.WriteObjectRequest_WriteObjectSpec{
+					WriteObjectSpec: &storagepb.WriteObjectSpec{
+						Resource: w.ObjectAttrs.toProtoObject(w.o.bucket),
+					},
+				}
+			}
+		}
+
+		err = w.stream.Send(req)
+		if err == nil {
+			return nil
+		} else if err != nil && err != io.EOF {
+			return err
+		}
+
+		// err was io.EOF. The client-side of a stream only gets an EOF on Send
+		// when the backend closes the stream and wants to return an error
+		// status. Closing the stream receives the status as an error.
+		_, err = w.stream.CloseAndRecv()
+
+		// Set first to true so that if the request is retried, the stream is
+		// reopened.
+		first = true
```

Comment:
This way of doing state management for the stream is a bit concerning. Does it make sense to have this retry loop continue if the stream has been closed with a retryable error, for example? Have you been able to test out locally whether this kind of retry actually works?

Reply:
I haven't tested retry, no, because I can't test the current retryable scenarios without a mock backend. I agree its a little precarious, but we need a way to 1. convey if the stream needs to be (re)opened 2. convey if the "first message" payload should be sent. Do you have another idea? Happy to switch it up.

Reply:
I've cleaned this up a lot since refactoring the retry logic. `first` is just determined on each loop iteration based on if `sent == 0`.